cloud/openstack/os_keystone_endpoint.py

```diff
@@ -0,0 +1,334 @@
+            changed = True
+        else:
+            if url and (enabled is not None
+                        and current_endpoint.enabled != enabled):
+                changed = True
+            if hasattr(current_endpoint, 'region') \
+                    and current_endpoint.region != region or \
+                    not hasattr(current_endpoint, 'region') \
+                    and region is not None:
+                changed = True
+            if url:
+                if current_endpoint.url != url:
+                    changed = True
+            else:
+                if hasattr(current_endpoint, 'internalurl') \
+                        and current_endpoint.internalurl != internal_url \
+                        or not hasattr(current_endpoint, 'internalurl') \
+                        and internal_url is not None:
+                    changed = True
+                if hasattr(current_endpoint, 'adminurl') \
+                        and current_endpoint.adminurl != admin_url \
+                        or not hasattr(current_endpoint, 'adminurl') \
+                        and admin_url is not None:
+                    changed = True
+                if hasattr(current_endpoint, 'publicurl') \
+                        and current_endpoint.publicurl != public_url \
+                        or not hasattr(current_endpoint, 'publicurl') \
+                        and public_url is not None:
+                    changed = True
+
+        # Handle check mode.
+        if module.check_mode:
+            if changed and state == "absent":
+                changed = False
+            module.exit_json(changed=changed)
+
+        args = {}
+        for endpoint in ["public_url", "internal_url", "admin_url"]:
+            args[endpoint] = module.params.get(endpoint)
+
+        # Do changes.
+        if changed:
+            if state == "present":
+                if current_endpoint is None:
+                    cloud.create_endpoint(service_id,
+                                          interface=interface,
+                                          url=url,
+                                          region=region,
+                                          enabled=enabled,
+                                          **args)
+                    changed = True
+                else:
+                    # Update endpoints.
+                    # Shade endpoint update is only possible in API V3.
+                    # So we can only use interface, url and region
+                    if url:
+                        cloud.update_endpoint(current_endpoint.id,
+                                              url=url,
+                                              interface=interface,
+                                              region=region)
+                    else:
+                        # No update possible in V2.0. So we must delete
+                        # and recreate.
+                        cloud.delete_endpoint(current_endpoint.id)
```

Comment:
This prevents the module from being idempotent. We cannot do this. Since this is a module that could only be used by cloud operators it is expected they should know if they are using v2.0 and v3, and the appropriate capabilities. Therefore, if shade returns and error letting them know the cloud doesnt support this feature, it should be passed to the user.